bin/ocamllsp.ml

```diff
@@ -0,0 +1,133 @@
+           compile %s is not known."
+          ocamllsp_exe_name
+          (Package_name.to_string compiler_package_name)
+          ocamllsp_exe_name
+      ]
+  ;;
+end
+
+(* Replace the current dune process with ocamllsp. *)
+let run_ocamllsp common ~args =
+  let exe_path_string = Path.to_string ocamllsp_exe_path in
+  Console.print_user_message
+    (Dune_rules.Pkg_build_progress.format_user_message
+       ~verb:"Running"
+       ~object_:
+         (User_message.command (String.concat ~sep:" " (ocamllsp_exe_name :: args))));
+  Console.finish ();
+  restore_cwd_and_execve common exe_path_string (exe_path_string :: args) Env.initial
+;;
+
+let build_ocamllsp common =
+  let open Fiber.O in
+  let+ result =
+    Build_cmd.run_build_system ~common ~request:(fun _build_system ->
+      Action_builder.path ocamllsp_exe_path)
+  in
+  match result with
+  | Error `Already_reported -> raise Dune_util.Report_error.Already_reported
+  | Ok () -> ()
+;;
+
+(* Ocamllsp needs to be compiled with the same version of the ocaml
+   compiler as the code it's analyzing, so look up the version of the
+   ocaml compiler in the lockdir so it can be added to the constraints
+   when solving ocamllsp's lockdir. *)
+let locked_ocaml_compiler_version () =
+  let open Memo.O in
+  let context =
+    (* Dev tools are only ever built with the default context. *)
+    Context_name.default
+  in
+  let* result = Dune_rules.Lock_dir.get context in
+  match result with
+  | Error _ -> Memo.return @@ Error Failure.Reason.No_lock_dir
+  | Ok { packages; _ } ->
+    (match Package_name.Map.find packages compiler_package_name with
+     | None ->
+       let+ lock_dir_path = Dune_rules.Lock_dir.get_path context >>| Option.value_exn in
+       Error (Failure.Reason.No_ocaml_lockfile_in_lock_dir { lock_dir_path })
+     | Some pkg -> Memo.return @@ Ok pkg.info.version)
+;;
+
+let is_in_dune_project builder =
+  Workspace_root.create
+    ~default_is_cwd:(Common.Builder.default_root_is_cwd builder)
+    ~specified_by_user:(Common.Builder.root builder)
+  |> Result.is_ok
+;;
+
+let term =
+  let+ builder = Common.Builder.term
+  and+ args = Arg.(value & pos_all string [] (info [] ~docv:"ARGS")) in
+  match is_in_dune_project builder with
+  | false -> raise (User_error.E (Failure.message Failure.Reason.Not_in_dune_project))
+  | true ->
+    let common, config = Common.init builder in
+    Scheduler.go ~common ~config (fun () ->
+      let open Fiber.O in
+      locked_ocaml_compiler_version ()
+      |> Memo.run
+      >>= function
+      | Error failure_reason -> raise (User_error.E (Failure.message failure_reason))
+      | Ok ocaml_compiler_version ->
+        let* () = Lock_dev_tool.lock_ocamllsp ~ocaml_compiler_version in
+        let+ () = build_ocamllsp common in
+        run_ocamllsp common ~args)
+;;
+
+let info =
+  let doc = "Run ocamllsp, installing it as a dev tool if necessary." in
+  Cmd.info "ocamllsp" ~doc
```

Comment:
```suggestion
  Cmd.info "lsp" ~doc
```

Small change but important to keep the DX consistent with `fmt`, `doc`, etc.

Reply:
The reason I think it should remain `dune ocamllsp` is this command is equivalent to running `ocamllsp`. `dune fmt` and `dune doc` both perform some higher-level function that inolves running `ocamlformat` and `odoc` respectively, but `dune ocamllsp` just builds `ocamllsp` and then runs it.

Reply:
For now, just make it `dune ocaml lsp`. That's where we put commands that aren't often useful to the user.

Reply:
@rgrinberg what do you think about `dune tools ocamllsp`? There was some discussion yesterday about introducing a `dune tools` command group for commands that just build and run some other tool.

Reply:
I guess that's fine too. Maybe it should be `dune tools exec ocamllsp`? I imagine something like `dune tools describe` being useful as well.

Reply:
I changed the command to `dune tools exec ocamllsp`.